R/check-cardinalities.R

```diff
@@ -78,7 +87,7 @@
 #'
 #' # This passes:
 #' check_cardinality_0_1(d1, a, d3, c)
-check_cardinality_0_n <- function(parent_table, pk_column, child_table, fk_column) {
+check_cardinality_0_n <- function(x, y, ...) {
```

Comment:
Just saw this now, sorry: actually the notation `0_n` refers only to the values in the child table. It should be read: 1 value in the primary table refers to 0 to n values in the child table. I was never very happy with the names of these functions, precisely because of this potential for misunderstanding... (but it's near impossible to improve, at least it seemed so in the past)

Reply:
I believe it is incorrect that "0 is x, n is y", my understanding is that `0_1` means that for `y`(child table), the unique key column(s) in `x` (parent) are matched 0 to 1 times.

* `check_cardinality_0_n` doesn't mean `check_cardinality_0_to_n` but means `check_cardinality_1_to_0_or_more`
* `check_cardinality_1_n` doesn't mean `check_cardinality_1_to_n` (at least with documented meaning of `n`, which is "more than 1") but means `check_cardinality_1_to_1_or_more`
* `check_cardinality_1_1` does mean `check_cardinality_1_to_1` by "accident"
* `check_cardinality_0_1` doesn't mean `check_cardinality_0_to_1` but means `check_cardinality_1_to_0_or_1`

I believe that many dm functions are more confusing than we might realise once we're used to it, so I personally wouldn't mind mentioning child and parent explicitly.

I remember when reviewing the original PR, that this was not easy to understand to me, and it wasn't today coming back to it, so maybe we should consider new names ?

* check_cardinality_1_to_01
* check_cardinality_1_to_0n
* check_cardinality_1_to_1n
* check_cardinality_1_to_1
(with `n` here meaning "1 or more") 

Or just : 

* check_1_to_01
* check_1_to_0n
* check_1_to_1n
* check_1_to_1

Having separate help files linked by `seeAlso` tags might make the doc less intimidating too, now it starts with and contains a lot of abstraction about the set of functions. I think I'd rather read a specific `check_1_to_0n` help file which would also make it easy ton infer what other functions do.

For this function I am not sure how it would work without `pk_column` and `fk_column` since we're outside of a dm so which columns to examine, if not explicit, could only be done AFAICT by name intersection or subsetting columns before feeding to the function. Could you clarify what you have in mind ?

@krlmlr @TSchiefer

Reply:
Right, `x` is always 1, `y` can be anything between 0 and n.

I went ahead and implemented. If `x_select` and `y_select` are not given, we compare the full data frames, matching columns by position. (Do we need to be more explicit about that in the docs?)

Once we think of a substantially better name, we can still supersede. For now these are as good as they get.